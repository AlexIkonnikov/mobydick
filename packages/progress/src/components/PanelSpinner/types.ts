interface IPanelSpinnerProps {
  isLoading: boolean;
  speed?: number;
  isError?: boolean;
  onCancel?: () => void;
}
export type PanelSpinnerProps = IPanelSpinnerProps;
