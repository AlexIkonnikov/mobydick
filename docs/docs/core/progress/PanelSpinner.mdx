import ReactPlayer from 'react-player';
import PanelSpinner from './assets/PanelSpinner.mov';

<ReactPlayer playing controls url={PanelSpinner} loop={true} />

Reusable PanelSpinner component

## Example

```tsx
import { PanelSpinner, View} from '@lad-tech/mobydick-core';

export const PanelSpinnerWidget = () => {
  return (
    <View>
      <PanelSpinner isLoading={true} />
      <PanelSpinner isLoading={false} />
      <PanelSpinner isLoading={true} onCancel={() => {}} />
    </View>
  );
};

```

## `Props`

### <div class="label required basic">Required</div>**`isLoading`**

| TYPE    |
|:--------|
| boolean |

Show loading animation

### `duration`

| TYPE   | DEFAULT |
|:-------|:--------|
| number | 2500    |

Animation duration in ms

### `isError`

| TYPE    |
|:--------|
| boolean |

Show error

### `onCancel`

| TYPE       |
|:-----------|
| () => void |

Callback to cancel animation

