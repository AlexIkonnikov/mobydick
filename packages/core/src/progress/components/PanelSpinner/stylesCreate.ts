import {createStyles} from '../../../styles';
import px from '../../../styles/utils/px';

const stylesCreate = createStyles(({spaces, colors}) => ({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    padding: spaces.Space20,
    backgroundColor: colors.BgPrimary,
    borderColor: colors.BorderSoft,
    borderWidth: px(0.5),
    borderRadius: px(200),
    shadowColor: colors.BgBlack,
    // ios
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.3,
    shadowRadius: px(12),
    // android
    elevation: 10,
  },
  insideView: {
    justifyContent: 'center',
    alignItems: 'center',
    width: spaces.Space48,
    aspectRatio: 1,
    borderRadius: spaces.Space24,
  },
}));
export default stylesCreate;
