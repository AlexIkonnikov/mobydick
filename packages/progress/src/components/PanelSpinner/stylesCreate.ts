import {IUseStylesTheme} from '@npm/mobydick-styles';
import {StyleSheet} from 'react-native';

const defaultStyle = (theme: IUseStylesTheme) =>
  StyleSheet.create({
    container: {
      justifyContent: 'center',
      alignItems: 'center',
      width: theme.spaces.Space64,
      aspectRatio: 1,
      backgroundColor: theme.colors.BgPrimary,
      borderRadius: 44,
      shadowColor: theme.colors.BgBlack,
      // ios
      shadowOffset: {
        width: 0,
        height: 10,
      },
      shadowOpacity: 1,
      shadowRadius: 16,
      // android
      elevation: 16,
    },
    errorView: {
      backgroundColor: theme.colors.IconMuted,
      width: 32,
      height: 32,
      borderRadius: 16,
    },
  });

const stylesCreate = (theme: IUseStylesTheme) => {
  return defaultStyle(theme);
};
export default stylesCreate;
