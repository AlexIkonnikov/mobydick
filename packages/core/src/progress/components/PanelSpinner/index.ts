import PanelSpinner from './PanelSpinner';

export * from './types';

export {PanelSpinner};
