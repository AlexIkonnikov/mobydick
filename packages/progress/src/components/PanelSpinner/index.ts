import PanelSpinner from './PanelSpinner';

export default PanelSpinner;
