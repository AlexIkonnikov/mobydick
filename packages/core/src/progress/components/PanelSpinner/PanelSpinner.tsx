import {FC} from 'react';
import {View} from 'react-native';

import SimpleIcon from '../../../styles/icons/font/SimpleIcon';
import useStyles from '../../../styles/hooks/useStyles';
import {Spinner} from '../Spinner';
import {ISizeSpinner} from '../Spinner/types';
import rem from '../../../styles/utils/rem';
import Pressable from '../../../basic/components/Pressable/Pressable';

import stylesCreate from './stylesCreate';
import {PanelSpinnerProps} from './types';

const PanelSpinner: FC<PanelSpinnerProps> = props => {
  const {isLoading, duration = 2500, isError = false, onCancel} = props;
  const [styles, theme] = useStyles(stylesCreate);

  const SpinnerStatus = () => {
    return isError ? (
      <View
        style={[
          styles.insideView,
          {backgroundColor: theme.colors.BgSecondary},
        ]}>
        <SimpleIcon
          name="icon-cancel"
          size={rem(32)}
          color={theme.colors.IconMuted}
        />
      </View>
    ) : (
      <View style={styles.insideView}>
        <SimpleIcon
          name="icon-check"
          size={rem(48)}
          color={theme.colors.IconBase}
        />
      </View>
    );
  };

  return (
    <View style={styles.container}>
      {isLoading ? (
        <>
          <Spinner duration={duration} size={ISizeSpinner.L} />
          {!!onCancel && (
            <Pressable
              style={[
                styles.insideView,
                {
                  position: 'absolute',
                },
              ]}
              onPress={onCancel}>
              <SimpleIcon name="icon-cancel" color={theme.colors.IconMuted} />
            </Pressable>
          )}
        </>
      ) : (
        <SpinnerStatus />
      )}
    </View>
  );
};

export default PanelSpinner;
