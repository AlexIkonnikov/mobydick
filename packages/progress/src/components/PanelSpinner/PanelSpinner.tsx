import {Pressable} from '@npm/mobydick-core';
import {SimpleIcon, useStyles} from '@npm/mobydick-styles';
import React, {FC} from 'react';
import {View} from 'react-native';

import Spinner from '../Spinner';
import {ISizeSpinner} from '../Spinner/types';

import stylesCreate from './stylesCreate';
import {PanelSpinnerProps} from './types';

const PanelSpinner: FC<PanelSpinnerProps> = props => {
  const {isLoading, speed = 2500, isError = false, onCancel} = props;
  const [styles, theme] = useStyles(stylesCreate);

  return (
    <View style={styles.container}>
      {isLoading ? (
        <>
          <Spinner speed={speed} size={ISizeSpinner.L} />
          {!!onCancel && (
            <Pressable
              style={[
                styles.insideView,
                {
                  position: 'absolute',
                },
              ]}
              onPress={onCancel}>
              <SimpleIcon name="icon-cancel" color={theme.colors.IconMuted} />
            </Pressable>
          )}
        </>
      ) : isError ? (
        <View
          style={[
            styles.insideView,
            {backgroundColor: theme.colors.BgSecondary},
          ]}>
          <SimpleIcon
            name="icon-cancel"
            size={32}
            color={theme.colors.IconMuted}
          />
        </View>
      ) : (
        <View style={styles.insideView}>
          <SimpleIcon
            name="icon-check"
            size={48}
            color={theme.colors.IconBase}
          />
        </View>
      )}
    </View>
  );
};

export default PanelSpinner;
