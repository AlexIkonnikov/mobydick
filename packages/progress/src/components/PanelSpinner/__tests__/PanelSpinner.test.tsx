import React from 'react';
import {render} from '@testing-library/react-native';

import PanelSpinner from '../index';

describe('@npm/mobydick-progress/PanelSpinner', () => {
  it('renders correctly', () => {
    const {toJSON} = render(<PanelSpinner isSpin />);
    expect(toJSON()).toMatchSnapshot();
  });
});
