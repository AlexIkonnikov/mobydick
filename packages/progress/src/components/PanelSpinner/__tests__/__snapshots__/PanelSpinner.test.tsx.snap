// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@npm/mobydick-progress/PanelSpinner renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1,
      "backgroundColor": "#FFF",
      "borderRadius": 44,
      "elevation": 16,
      "justifyContent": "center",
      "shadowColor": "#20242D",
      "shadowOffset": Object {
        "height": 10,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 16,
      "width": 64,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "transform": Array [
          Object {
            "rotate": "0deg",
          },
        ],
      }
    }
  >
    <SvgMock
      fill="#2B78EE"
    />
  </View>
</View>
`;

exports[`@npm/mobydick-progress/PanelSpinner renders correctly canceled 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1,
      "backgroundColor": "#FFF",
      "borderRadius": 44,
      "elevation": 16,
      "justifyContent": "center",
      "shadowColor": "#20242D",
      "shadowOffset": Object {
        "height": 10,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 16,
      "width": 64,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "transform": Array [
          Object {
            "rotate": "0deg",
          },
        ],
      }
    }
  >
    <SvgMock
      fill="#2B78EE"
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#9BA1B0",
          "borderRadius": 16,
          "height": 32,
          "width": 32,
        },
        Object {
          "backgroundColor": "green",
          "position": "absolute",
        },
      ]
    }
  />
</View>
`;

exports[`@npm/mobydick-progress/PanelSpinner renders correctly done 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1,
      "backgroundColor": "#FFF",
      "borderRadius": 44,
      "elevation": 16,
      "justifyContent": "center",
      "shadowColor": "#20242D",
      "shadowOffset": Object {
        "height": 10,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 16,
      "width": 64,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#9BA1B0",
          "borderRadius": 16,
          "height": 32,
          "width": 32,
        },
        Object {
          "backgroundColor": "red",
        },
      ]
    }
  />
</View>
`;

exports[`@npm/mobydick-progress/PanelSpinner renders correctly error 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "aspectRatio": 1,
      "backgroundColor": "#FFF",
      "borderRadius": 44,
      "elevation": 16,
      "justifyContent": "center",
      "shadowColor": "#20242D",
      "shadowOffset": Object {
        "height": 10,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 16,
      "width": 64,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#9BA1B0",
        "borderRadius": 16,
        "height": 32,
        "width": 32,
      }
    }
  />
</View>
`;
